### DISCUSSION TITLE
Skip all unrecognized chars

### DISCUSSION THREAD

**Discussion by ccleve** [pest-parser/pest#727] [Q&A] [answered]
How do I write a grammar that skips all unrecognized characters that are between tokens?

For example, consider this input string: "##abc& &def%%"

I would like to recognized "abc" and "def" as words, and skip everything else. This is what I have tried:

```
main = { SOI ~ word+ ~ EOI }
word = { 'a'..'z'}
WHITESPACE = { ANY+? }
```

```
main = { SOI ~ (other ~ word)+ ~ other ~ EOI }
word = { 'a'..'z'}
other = { ANY*? }
```

```
main = { SOI ~ (other ~ word)+ ~ other ~ EOI }
word = { ('a'..'z')+ }
other = { !(word) }

```

```
main = { SOI ~ other*? ~ (word ~ other*?)+ ~ EOI }
word = { 'a'..'z'}
other = { ANY }

```

The problem with the first two is that WHITESPACE and "other" recognize too much. Essentially, I need a reluctant quantifier. In a regular expression "?" does the trick, but it's not working as expected here.

I don't know what the issue is with the last two. The very last one really should work, but it is expecting another word at the end and I don't know why.

**Comment by xiaket** [comment, marked as answer]
I believe the following peg will work:

```
main = { SOI ~ words ~ EOI }
words = { unwanted* ~ word ~ unwanted* ~ words? ~ unwanted* }
word = { ('a'..'z')+ }
unwanted = _{
    !(word)
    ~ ANY
}
```
